packages/world/src/modules/keysintable/KeysInTableHook.sol

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+import { IStoreHook } from "@latticexyz/store/src/IStore.sol";
+
+import { ResourceSelector } from "../../ResourceSelector.sol";
+
+import { MODULE_NAMESPACE, KEYS_LENGTH_NAMESPACE, USED_KEYS_NAMESPACE } from "./constants.sol";
+import { KeysInTable } from "./tables/KeysInTable.sol";
+import { KeysInTableLength } from "./tables/KeysInTableLength.sol";
+import { UsedKeysIndex, UsedKeysIndexData } from "./tables/UsedKeysIndex.sol";
+import { ArrayLib } from "../utils/ArrayLib.sol";
+import { getTargetTableSelector } from "../utils/getTargetTableSelector.sol";
+
+/**
+ * This is a very naive and inefficient implementation for now.
+ * We can optimize this by adding support for `setIndexOfField` in Store
+ * and then replicate logic from solecs's Set.sol.
+ * (See https://github.com/latticexyz/mud/issues/444)
+ *
+ * Note: if a table with composite keys is used, only the first key is indexed
+ */
+contract KeysInTableHook is IStoreHook {
+  using ArrayLib for bytes32[];
+  using ResourceSelector for bytes32;
+
+  function handleSet(bytes32 tableId, bytes32[] memory key) internal {
+    bytes32 keysInTableTableId = getTargetTableSelector(MODULE_NAMESPACE, tableId);
+    bytes32 keysInTableLengthTableId = getTargetTableSelector(KEYS_LENGTH_NAMESPACE, tableId);
```

Comment:
instead of using a different singleton table per source table to index, why don't we use a single `KeysInTableTable` with a `{ sourceTable: "bytes32" }` primary key?

Reply:
I copied this pattern from `KeysWithValue`. Is there any reason for that module to do it differently?

Reply:
> I copied this pattern from KeysWithValue. Is there any reason for that module to do it differently?

context from discord: it probably should be a single table for `KeysWithValue` too.

> In that case, they have a singleton table of the values in each source table, but it could be { sourceTable: "bytes32", keyHash: "bytes32" } 

You mean `{ sourceTable: "bytes32", valueHash: "bytes32" }` as the primary key for the table in `KeysWithValue`? If so, I agree!